Core/AI: implement AI function DelayedCast to delay a spell cast

gerripeach opened the pull request:
**Target branch(es):** 3.3.5/master

- [X] 3.3.5
- [?] master - has to be ported & maybe fix compile due to slightly different structure

**Tests performed:** (Does it build, tested in-game, etc.)
does build
tested with several creatues

joschiwald commented on src/server/game/AI/CoreAI/UnitAI.cpp:
```diff
@@ -193,6 +193,35 @@ void UnitAI::DoCastAOE(uint32 spellId, bool triggered)
+            if (_aliveCheck && _caster->IsAlive())
+                _caster->CastSpell(_target, _spellId, _triggered);
+            else
+                _caster->CastSpell(_target, _spellId, _triggered);
```
nice copy paste from somewhere

gerripeach replied:
the idea from DoCastDelayed is not from me. Tell me the real author and I change the commit message

Shauren replied:
Thats not what he is talking about - look closely at the code. If branch is exactly the same as else branch

gerripeach replied:
I don't see the problem. My trinity core branch is 3(?) years behind. So there might be some issues with different code

Seyden replied:
Just look closely dude.. your condition doesn't make any sense

Shauren replied:
You don't understand, I'm not talking about git branches but this https://en.wikipedia.org/wiki/Branch_(computer_science)

gerripeach replied:
ty: shauren

joschiwald commented on src/server/game/AI/CoreAI/UnitAI.cpp:
```diff
@@ -193,6 +193,35 @@ void UnitAI::DoCastAOE(uint32 spellId, bool triggered)
+        }
+
+        Unit *_caster;
+        Unit *_target;
```
this leads to crashes
btw. since we have a taskscheduler i don't see any use for this function

Krudor replied:
Because TaskScheduler wouldn't work if its not being updated from UpdateAI?

Treeston replied:
`m_Events` still processes while dead. Why is this not sufficient?

Krudor replied:
Sure this will suffice, but doesn't it feel hacky?
This would only allow us to cast spells with a certain delay, that aren't affected by death for example.

How about when we need to do more than just cast a delayed spell upon death.
I don't have a suggestion right now, but I don't see why we shouldn't be able to schedule after-death events properly, even though in 99% of the cases it's not relevant.

Treeston replied:
You seem to be misunderstanding - I'm not agreeing with the approach of the PR.

I'm trying to figure out why (or whether) you think the current state of affairs - using custom `Event`s for the very few edge cases where this is relevant and no other master AI is around to handle it cleanly - is insufficient.

Krudor replied:
I think using custom Events for special cases like this is fine.

jackpoz replied:
who won here ? is it still useful to have this event ?

Treeston replied:
Doesn't seem like there's any convincing argument for it. Closing.

Shauren reviewed (changes requested):
fix the if

Aokromes commented:
https://github.com/TrinityCore/TrinityCore/pull/20795/files#diff-ab7e4a0d883137967b73ba1fd4e3c241R205 he changed the if days ago.

ccrs commented:
it still makes no sense

Keader commented:
any example where it can be used?

gerripeach commented:
If you have any dead creature and this creatures has to cast after a specific time a spell nothing that uses the update can be used

Krudor commented:
I've encountered a similiar issue with this before, where Firelands bosses cast the legendary aura that allows players to drain their essence, a few seconds after their death.

I would either fix the underlying issue with after-death events (UpdateAI not triggering right?), or work around it with a BasicEvent like you're doing here, just don't make it generic.

This is more like a generic hackfix to circumvent the real issue.

Seyden commented:
Bro.. this if clause is always shitty.. don't try to bypass your problem with an else if..

try it with only ONE if condition.. you can use multiple condition in 1 if clause

Commit:
Core/AI: Implement AI function to cast a spell DelayedCast

useful if want to schedule a spell cast and don't want to use timer(less messy code in scripts)

Treeston commented:
> for exaple have a look at chogall. this boss summons adds. when these adds die they will casted 4 sec after death 2 spells. *I could either call a method on chogall and let him do the update* or use my method

Yeah, so what's the point of this?

PS: Are you even sure the creatures are fully dead in your example? Sounds like something Blizzard would use feign death for, dead creatures practically never cast spells.

Krudor commented:
> PS: Are you even sure the creatures are fully dead in your example? Sounds like something Blizzard would use feign death for, dead creatures practically never cast spells.

I'm going to assume you did not read my reply, and this I'll repeat it again down below.

I know for a fact Firelands bosses cast spells after they have died, to give you an example.

We need SOMETHING to support this, but as I mentioned, THIS implementation feels hacky and more like a workaround.

Treeston commented:
Calling `UpdateAI` for dead creatures is a lot of extra jumps for very few special cases, I'm not a big fan.

Is there some issue with using `Event`s?

Killyana commented:
In some cases like https://github.com/TrinityCore/TrinityCore/issues/20591 we need to despawn the dead npc on spell hit, r00ty-tc already did an initial work https://github.com/r00ty-tc/TrinityCore/commit/74f640f4c348415686c8ff7ef8c6992e3b789994
